File: python/src/multiagent_boilerplate/tools/builtin/mcp_call.py
from __future__ import annotations

from ...mcp_.client import McpClientManager
from ..types import ToolContext


class McpCallTool:
    name = "mcp_call"
    description = (
        "Call a tool exposed by a connected external MCP server (config/mcp_servers.json), by server name and tool."
    )
    input_schema = {
        "type": "object",
        "required": ["server", "tool", "input"],
        "properties": {"server": {"type": "string"}, "tool": {"type": "string"}, "input": {"type": "object"}},
    }
    exposable = False

    def __init__(self, manager: McpClientManager) -> None:
        self._manager = manager

    async def execute(self, input_: dict, ctx: ToolContext) -> object:
        return await self._manager.call_tool(input_["server"], input_["tool"], input_["input"])

File: typescript/src/tools/builtin/mcpCall.ts
import type { McpClientManager } from "../../mcp/client.js";
import type { Tool } from "../types.js";

export function createMcpCallTool(manager: McpClientManager): Tool<{ server: string; tool: string; input: Record<string, unknown> }, unknown> {
  return {
    name: "mcp_call",
    description: "Call a tool exposed by a connected external MCP server (config/mcp-servers.json), by server name and tool name.",
    inputSchema: {
      type: "object",
      required: ["server", "tool", "input"],
      properties: {
        server: { type: "string" },
        tool: { type: "string" },
        input: { type: "object" },
      },
    },
    async execute(input) {
      return manager.callTool(input.server, input.tool, input.input);
    },
  };
}
